finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/Stream.scala

```diff
@@ -153,7 +153,12 @@ object Stream {
     override def onCancel: Future[Reset] = cancelP
 
     private[this] val endOnReleaseIfEnd: Try[Frame] => Unit = {
-      case Return(f) => if (f.isEnd) endP.become(f.onRelease)
+      case Return(f) =>
```

Comment:
This change is not technically required to fix the above issue.  However, when this issue is triggered this line causes exceptions to be thrown because `endP` becomes satisfied by cancel and `endP.become` causes `f.onRelease` to be satisfied.  This leads to `f.onRelease` being set again when `f.release()` is called, which is an illegal operation.

We change this code slightly to protect against such an exception.  We only allow updates to flow from `f.onRelease` to `endP` and not the other way around.